### PR TITLE
Add logging for conda/mamba install

### PR DESCRIPTION
**Pull request by janjagusch**
With this PR you are now able to see the stdout and stderr of the conda/mamba install in real time in your terminal.

We've noticed that when running `conda-lock install` on large lock files, you to stare at a blank terminal for quite a long time. We thought it'd be nice to see what is currently happening within conda/mamba.

The commands `conda-lock lock` and `conda-lock install` now have an additional `--log-level` argument with the default set to `INFO`. The general conda/mamba stdout is logged as `INFO` and the stderr as `ERROR`.

One nice side effect was that we could get rid of the rather complicated `_handle_subprocess_stdout` function, because the errors are already nicely logged.

---

Here are some examples of how `conda-lock install` looks with logging:

```
INFO:root:
INFO:root:                  __    __    __    __
INFO:root:                 /  \  /  \  /  \  /  \
INFO:root:                /    \/    \/    \/    \
INFO:root:███████████████/  /██/  /██/  /██/  /████████████████████████
INFO:root:              /  / \   / \   / \   / \  \____
INFO:root:             /  /   \_/   \_/   \_/   \    o \__,
INFO:root:            / _/                       \_____/  `
INFO:root:            |/
INFO:root:        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
INFO:root:        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
INFO:root:        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
INFO:root:        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
INFO:root:        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
INFO:root:        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝
INFO:root:
INFO:root:        mamba (0.7.12) supported by @QuantStack
INFO:root:
INFO:root:        GitHub:  https://github.com/mamba-org/mamba
INFO:root:        Twitter: https://twitter.com/QuantStack
INFO:root:
INFO:root:█████████████████████████████████████████████████████████████
INFO:root:
INFO:root:Preparing transaction: ...working... done
INFO:root:Verifying transaction: ...working... done
INFO:root:Executing transaction: ...working... done
```

And also one example for when an error occurs:

```
INFO:root:
INFO:root:                  __    __    __    __
INFO:root:                 /  \  /  \  /  \  /  \
INFO:root:                /    \/    \/    \/    \
INFO:root:███████████████/  /██/  /██/  /██/  /████████████████████████
INFO:root:              /  / \   / \   / \   / \  \____
INFO:root:             /  /   \_/   \_/   \_/   \    o \__,
INFO:root:            / _/                       \_____/  `
INFO:root:            |/
INFO:root:        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
INFO:root:        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
INFO:root:        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
INFO:root:        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
INFO:root:        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
INFO:root:        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝
INFO:root:
INFO:root:        mamba (0.7.12) supported by @QuantStack
INFO:root:
INFO:root:        GitHub:  https://github.com/mamba-org/mamba
INFO:root:        Twitter: https://twitter.com/QuantStack
INFO:root:
INFO:root:█████████████████████████████████████████████████████████████
INFO:root:
INFO:root:
INFO:root:Downloading and Extracting Packages
INFO:root:
INFO:root:mypackage |            |   0%
INFO:root:mypackage |            |   0%
ERROR:root:
ERROR:root:CondaHTTPError: HTTP 401 UNAUTHORIZED for url <https://mycondaserver.cloud/mychannel/mypackage>
ERROR:root:Elapsed: 00:00.180086
ERROR:root:
ERROR:root:An HTTP error occurred when trying to retrieve this URL.
ERROR:root:HTTP errors are often intermittent, and a simple retry will get you on your way.
ERROR:root:
```

---

I wanted to make this change as minimal as possible, but if you're okay with it I would propose some more changes:

- Creating a `Logger` class instead of using the default logger
- Converting all `print` into `logging.info`
- Formatting the logging to remove the log level from the output

---

Note: The code for streaming from the subprocess in taken from [here](https://stackoverflow.com/a/17698359/7380270).

### CODE REVIEW

**Comment by mariusvniekerk** [comment]
Yeah, we probably need to pass some additional flags so that things don't try too hard to have nice updating progressbars etc

**Comment by janjagusch** [comment]
> Yeah, we probably need to pass some additional flags so that things don't try too hard to have nice updating progressbars etc

Sorry for the late response, I've been quite busy the week after easter. I agree that progress bars might be an issue. For example, you can see that currently progress bar calls are logged twice:

```
INFO:root:Downloading and Extracting Packages
INFO:root:
INFO:root:data_access_python-0 |            |   0%
INFO:root:data_access_python-0 |            |   0%
```

But to be honest, I have no idea how to fix this. I also couldn't find much on the topic besides SO posts that are a decade old. Do you know what might be the right direction here?

**Comment by mariusvniekerk** [comment]
I think `--quiet` is probably good enough for us

File: conda_lock/conda_lock.py [hunk 1]
```diff
@@ -148,6 +148,20 @@ def print_proc(proc):
...
+    for logline in stdout:
+        if logline.endswith("\n"):
+            logline = logline[:-1]
+        if "%" in logline:
+            logline = logline.split()[0]
+            if logline not in cache:
+                yield logline
+                cache.add(logline)
+        else:
+            yield logline
```
Comment (janjagusch) [inline conda_lock/conda_lock.py:162]: i could clean this snippet up a bit more, but before i thought i'd ask you what you think of the overall approach.

File: conda_lock/conda_lock.py [hunk 2]
```diff
@@ -173,42 +186,29 @@ def do_conda_install(conda: PathLike, prefix: str, name: str, file: str) -> None
...
+        bufsize=1,
+        universal_newlines=True,
+    ) as p:
+        if p.stdout:
+            for line in _process_stdout(p.stdout):
+                logging.info(line)
+
+        if p.stderr:
+            for line in p.stderr:
+                logging.error(line if not line.endswith("\n") else line[:-1])
```
Comment (mariusvniekerk) [inline conda_lock/conda_lock.py:204]: can probably just do?

```suggestion
                logging.error(line.rstrip())
```

File: conda_lock/conda_lock.py (hunk 1 above)
Comment (mariusvniekerk) [reply conda_lock/conda_lock.py:162]: guessing this is an attempt to reduce progressbar spam logging.  Probably with --quiet we might not need this?

**Comment by janjagusch** [comment]
Okay, so I tried out a couple of things and didn't really find a solution that I'm 100% happy with:

- Option 1 using `--quiet`: Can be implemented very easily but it is ... well ... very quiet. The help is a bit confusing there. It does not only remove the progress bar, it removes any kind of progress. All of the output happens pretty much at the very end, so for most of the execution you're still blind.
- Option 2 using something custom: The output is nicer and more meaningful but it requires transforming the log lines *a lot*. It works well now, but I'm not sure what happens if conda changes its log output. I wouldn't want my `conda-lock install` to fail because of some error during the log parsing.

I put the two solutions in separate commits. I also added a little test suite, so you can try it out yourself if you want.

Just let me know what approach you like better and then i would go ahead and clean up and get everything ready for a final review. @mariusvniekerk 

---

Commit `f85a652d77b58b18ceae9a1dc7f6898f531bb7ad`:

```
docker run conda-lock-test:latest
INFO:root:
INFO:root:                  __    __    __    __
INFO:root:                 /  \  /  \  /  \  /  \
INFO:root:                /    \/    \/    \/    \
INFO:root:███████████████/  /██/  /██/  /██/  /████████████████████████
INFO:root:              /  / \   / \   / \   / \  \____
INFO:root:             /  /   \_/   \_/   \_/   \    o \__,
INFO:root:            / _/                       \_____/  `
INFO:root:            |/
INFO:root:        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
INFO:root:        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
INFO:root:        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
INFO:root:        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
INFO:root:        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
INFO:root:        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝
INFO:root:
INFO:root:        mamba (0.7.6) supported by @QuantStack
INFO:root:
INFO:root:        GitHub:  https://github.com/mamba-org/mamba
INFO:root:        Twitter: https://twitter.com/QuantStack
INFO:root:
INFO:root:█████████████████████████████████████████████████████████████
INFO:root:
INFO:root:
INFO:root:Downloading and Extracting Packages
INFO:root:
INFO:root:_libgcc_mutex-0.1
INFO:root:
INFO:root:ca-certificates-2020
INFO:root:
INFO:root:ld_impl_linux-64-2.3
INFO:root:
INFO:root:libstdcxx-ng-9.3.0
INFO:root:
INFO:root:libgomp-9.3.0
INFO:root:
INFO:root:_openmp_mutex-4.5
INFO:root:
INFO:root:libgcc-ng-9.3.0
INFO:root:
INFO:root:libffi-3.3
INFO:root:
INFO:root:libsodium-1.0.18
INFO:root:
INFO:root:libuuid-2.32.1
INFO:root:
INFO:root:ncurses-6.2
INFO:root:
INFO:root:xz-5.2.5
INFO:root:
INFO:root:zlib-1.2.11
INFO:root:
INFO:root:readline-8.0
INFO:root:
INFO:root:tk-8.6.10
INFO:root:
INFO:root:zeromq-4.3.4
INFO:root:
INFO:root:sqlite-3.35.4
INFO:root:
INFO:root:xeus-1.0.1
INFO:root:
INFO:root:python-3.8.8
INFO:root:
INFO:root:backcall-0.2.0
INFO:root:
INFO:root:backports-1.0
INFO:root:
INFO:root:decorator-5.0.6
INFO:root:
INFO:root:ipython_genutils-0.2
INFO:root:
INFO:root:nest-asyncio-1.5.1
INFO:root:
INFO:root:parso-0.8.2
INFO:root:
INFO:root:ptyprocess-0.7.0
INFO:root:
INFO:root:python_abi-3.8
INFO:root:
INFO:root:six-1.15.0
INFO:root:
INFO:root:debugpy-1.2.1
INFO:root:
INFO:root:jedi-0.18.0
INFO:root:
INFO:root:pexpect-4.8.0
INFO:root:
INFO:root:pickleshare-0.7.5
INFO:root:
INFO:root:python-dateutil-2.8.
INFO:root:
INFO:root:pyzmq-22.0.3
INFO:root:
INFO:root:tornado-6.1
INFO:root:
INFO:root:traitlets-5.0.5
INFO:root:
INFO:root:jupyter_core-4.7.1
INFO:root:
INFO:root:setuptools-49.6.0
INFO:root:
INFO:root:backports.functools_
INFO:root:
INFO:root:jupyter_client-6.2.0
INFO:root:
INFO:root:pygments-2.8.1
INFO:root:
INFO:root:wcwidth-0.2.5
INFO:root:
INFO:root:prompt-toolkit-3.0.1
INFO:root:
INFO:root:ipython-7.22.0
INFO:root:
INFO:root:ipdb-0.13.4
INFO:root:
INFO:root:ipykernel-5.5.3
INFO:root:
INFO:root:xeus-python-0.12.3
INFO:root:Preparing transaction: ...working... done
INFO:root:Verifying transaction: ...working... done
INFO:root:Executing transaction: ...working... done
```

Commit `9b3d7beb4e3a391a4550f4442d674ea64880f827`:

```
docker run conda-lock-test:latest
INFO:root:Preparing transaction: ...working... done
INFO:root:Verifying transaction: ...working... done
INFO:root:Executing transaction: ...working... done
```

File: conda_lock/conda_lock.py (hunk 2 above)
Comment (janjagusch) [reply conda_lock/conda_lock.py:204]: i actually didn't know about this, thanks!

File: conda_lock/conda_lock.py (hunk 1 above)
Comment (janjagusch) [reply conda_lock/conda_lock.py:162]: See https://github.com/conda-incubator/conda-lock/pull/91#issuecomment-818722333

**Comment by mariusvniekerk** [comment]
@janjagusch I think we can additionally also prune out the logging messages that are empty lines.  Sure that will remove some of the bits from the mamba header etc, but ultimately makes a nicer looking log?

**Comment by janjagusch** [comment]
> @janjagusch I think we can additionally also prune out the logging messages that are empty lines. Sure that will remove some of the bits from the mamba header etc, but ultimately makes a nicer looking log?
> 
> Agree with you that quiet is probably a bit too quiet by default.

I like the idea of removing the empty lines. I guess we could even build it such that the stripping starts after the mamba logo has been displayed. I might be able to get this done tomorrow, but next week sounds more realistic. :)

**Comment by janjagusch** [comment]
With https://github.com/conda-incubator/conda-lock/pull/91/commits/5e4df2c7de78b9f3aecd59246c976c222a1f8fe4, we are now also stripping leading empty lines.

I think this is ready for another review, @mariusvniekerk. :)

mamba:

```
INFO:root:                  __    __    __    __
INFO:root:                 /  \  /  \  /  \  /  \
INFO:root:                /    \/    \/    \/    \
INFO:root:███████████████/  /██/  /██/  /██/  /████████████████████████
INFO:root:              /  / \   / \   / \   / \  \____
INFO:root:             /  /   \_/   \_/   \_/   \    o \__,
INFO:root:            / _/                       \_____/  `
INFO:root:            |/
INFO:root:        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
INFO:root:        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
INFO:root:        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
INFO:root:        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
INFO:root:        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
INFO:root:        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝
INFO:root:
INFO:root:        mamba (0.7.12) supported by @QuantStack
INFO:root:
INFO:root:        GitHub:  https://github.com/mamba-org/mamba
INFO:root:        Twitter: https://twitter.com/QuantStack
INFO:root:
INFO:root:█████████████████████████████████████████████████████████████
INFO:root:
INFO:root:
INFO:root:Downloading and Extracting Packages
INFO:root:libgfortran5-9.3.0
INFO:root:libgfortran-5.0.0
INFO:root:libopenblas-0.3.12
INFO:root:libblas-3.9.0
INFO:root:python_abi-3.8
INFO:root:libcblas-3.9.0
INFO:root:liblapack-3.9.0
INFO:root:Preparing transaction: ...working... done
INFO:root:Verifying transaction: ...working... done
INFO:root:Executing transaction: ...working... done
```

conda:

```
INFO:root:Downloading and Extracting Packages
INFO:root:libgfortran5-9.3.0
INFO:root:libgfortran-5.0.0
INFO:root:libopenblas-0.3.12
INFO:root:libblas-3.9.0
INFO:root:python_abi-3.8
INFO:root:libcblas-3.9.0
INFO:root:liblapack-3.9.0
INFO:root:Preparing transaction: ...working... done
INFO:root:Verifying transaction: ...working... done
INFO:root:Executing transaction: ...working... done
```

**Pull request merged**